flink-queryable-state/flink-queryable-state-runtime/src/test/java/org/apache/flink/queryablestate/itcases/AbstractQueryableStateTestBase.java

```diff
@@ -361,6 +372,65 @@ public Integer getKey(Tuple2<Integer, Long> value) {
         }
     }
 
+    /** This test checks if custom Kryo serializers are loaded with correct classloader. */
+    @Test
+    public void testCustomKryoSerializerHandling() throws Exception {
+        final Deadline deadline = Deadline.now().plus(TEST_TIMEOUT);
+        final long numElements = 1;
+        final String stateName = "teriberka";
+
+        final String customSerializerClassName = "CustomKryo";
+        final URLClassLoader userClassLoader =
+                createLoaderWithCustomKryoSerializer(customSerializerClassName);
+
+        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+        env.setStateBackend(stateBackend);
+        env.setParallelism(maxParallelism);
+        // Very important, because cluster is shared between tests and we
+        // don't explicitly check that all slots are available before
+        // submitting.
+        env.setRestartStrategy(RestartStrategies.fixedDelayRestart(Integer.MAX_VALUE, 1000L));
+
+        // Custom serializer is not needed, it's used just to check if serialization works.
+        @SuppressWarnings("unchecked")
+        Class<Serializer<?>> customSerializerClass =
+                (Class<Serializer<?>>) userClassLoader.loadClass(customSerializerClassName);
+        env.getConfig().addDefaultKryoSerializer(Byte.class, customSerializerClass);
```

Comment:
I think we have to register a serializer instance to make this test work. If we only use a `Class`, then the system won't try to duplicate it.

Reply:
You're right, thanks for spotting this, I think I made one cleanup too much :)